packages/shared/src/components/plus/PlusIOS.tsx

```diff
@@ -0,0 +1,195 @@
+  const { displayToast } = useToastNotification();
+  const { user } = useAuthContext();
+  const productIds = useFeature(featureIAPProducts);
+  const [selectedOption, setSelectedOption] = useState<string | null>(null);
+
+  const selectionChange: OpenCheckoutFn = useCallback(({ priceId }) => {
+    setSelectedOption(priceId);
+  }, []);
+
+  const onContinue = useCallback(() => {
+    sendMessage(WebKitMessageHandlers.IAPSubscriptionRequest, {
+      productId: selectedOption,
+      appAccountToken: user?.subscriptionFlags?.appAccountToken,
+    });
+  }, [selectedOption, user?.subscriptionFlags?.appAccountToken]);
+
+  const productList = useMemo(() => Object.keys(productIds), [productIds]);
+
+  const { data: productOptions, ...query } = useQuery({
+    queryKey: ['iap-products'],
+    queryFn: async () => {
+      const products = promisifyEventListener(
+        'iap-products-result',
+        (event) => {
+          const productsRaw = !isNullOrUndefined(event?.detail)
+            ? JSON.parse(event.detail)
+            : [];
+
+          return productsRaw
+            ?.map((product: IAPProduct) => {
+              const duration = productIds[
+                product.attributes.offerName
+              ] as PlusPriceType;
+              const currencySymbol =
+                product.attributes.offers[0].priceFormatted.replace(
+                  /\d|\.|\s|,/g,
+                  '',
+                );
+
+              return {
+                label: product.attributes.name,
+                value: product.attributes.offerName,
+                price: {
+                  amount: product.attributes.offers[0].price,
+                  formatted: product.attributes.offers[0].priceFormatted,
+                  monthlyAmount: product.attributes.offers[0].price,
+                  monthlyFormatted: product.attributes.offers[0].priceFormatted,
+                },
+                currencyCode: product.attributes.offers[0].currencyCode,
+                currencySymbol,
+                extraLabel: product.attributes?.description?.standard,
+                appsId:
+                  product.attributes.offerName === 'annualSpecial'
+                    ? PlusPriceTypeAppsId.EarlyAdopter
+                    : PlusPriceTypeAppsId.Default,
+                duration,
+              };
+            })
+            .sort((a: { value: string }, b: { value: string }) => {
+              // Make sure that the products are sorted in the same order as the product list
+              // because the native code does not guarantee the order of the products
+              const aIndex = productList.indexOf(a.value);
+              const bIndex = productList.indexOf(b.value);
+              return aIndex - bIndex;
+            });
+        },
+      );
+
+      sendMessage(WebKitMessageHandlers.IAPProductList, productList);
+
+      return products;
+    },
+  });
+
+  useEffect(() => {
+    promisifyEventListener('iap-purchase-result', (event) => {
```

Comment:
Realizing this now but technically this should return `removeEventListener` so we can cleanup as well, something like:
```suggestion
    const unsubscribe = promisifyEventListener('iap-purchase-result', (event) => {
```

Reply:
:+1:

Reply:
I'll do that in a separate PR, as I'll have to fix all the others too. `promisifyEventListener` returns a Promise, so I'll need to update it to return the promise and a unsubscribe function.

In the meantime, I'll just use `globalThis.removeEventListener`

Reply:
yeah not blocking

Reply:
good catch!! i'm a great tech debt generator